pensieve/analysis.py

```diff
@@ -160,10 +167,93 @@ def run(self, current_date: datetime, dry_run: bool):
             return
 
         self.logger.info("Executing query for %s", self.experiment.slug)
-        self.bigquery.execute(sql, res_table_name)
+        result = self.bigquery.execute(sql, res_table_name)
         self._publish_view("week")
         self.logger.info("Finished running query for %s", self.experiment.slug)
 
+        return res_table_name
+
+    def _calculate_statistics(self, result_table: str):
+        """
+        Run statistics on metrics.
+        """
+
+        statistics_results = []
+
+        metrics_data = self.bigquery.table_to_dataframe(result_table)
+
+        for statistic in self.STANDARD_STATISTICS:
+            result_dict = {}
+            result_dict["statistic"] = statistic.name
+
+            # calculate statistics for specified branches
+            if statistic.branches is not None:
+                results_per_branch = metrics_data.groupby("branch")
+
+                for branch in statistic.branches:
+                    data = results_per_branch.get_group(branch)
+
+                    for metric in statistic.metrics:
+                        if metric in data:
+                            key_value_results = []
+
+                            for key, value in statistic.function(data[metric]).to_dict().items():
+                                key_value_results.append({"key": key, "value": value})
+
+                            statistics_results.append(
+                                {
+                                    "name": statistic.name,
+                                    "branch": branch,
+                                    "map_key_value": key_value_results,
+                                }
+                            )
+            else:
+                # otherwise pass entire dataframe to statistics function
+                key_value_results = []
+
+                for key, value in statistic.function(metrics_data).to_dict().items():
+                    key_value_results.append({"key": key, "value": value})
+
+                statistics_results.append(
+                    {"name": statistic.name, "branch": None, "map_key_value": key_value_results}
+                )
+
+        df_statistics_results = pandas.DataFrame.from_dict(statistics_results)
+
+        print(df_statistics_results)
+
+        table_id = f"{self.project}.{self.dataset}.statistics_{result_table}"
+
+        job = self.bigquery.client.load_table_from_dataframe(df_statistics_results, table_id)
```

Comment:
Currently, uploading [STRUCT / RECORD fields from load_table_from_dataframe] is only supported with `pyarrow 0.17.0` which is not published on pip, yet (but should be soon).

Reply:
We should put together a proposal for what these tables should look like.